README.md
# TYPO3 Extension ``google_for_jobs``

![google for Jobs](https://github.com/pegasuswerbeagentur/google_for_jobs/blob/master/Resources/Public/Icons/git-repo-teaser.JPG)

## 1. Features

- easy integration
- decide to hide/display fluidtemplate in frontend
- render structured data to your job posting
- more features coming soon

## 2. Usage

### 1) Installation

#### Installation using Composer

The recommended way to install the extension is by using [Composer][1]. In your Composer based TYPO3 project root, just do `composer require pegasus/google-for-jobs`. 

#### Installation as extension from TYPO3 Extension Repository (TER)

Download and install the extension with the extension manager module.

### 2) Setup

1) Include the static TypoScript of the extension.
2) Create some location and job records on a sysfolder.
3) Create a plugin on a page and select at least the sysfolder as startingpoint.

### 3) Sitemap

Add the following code to your setup.typoscript
```
plugin.tx_seo.config {
  xmlSitemap {
    sitemaps {
      jobs {
        provider = TYPO3\CMS\Seo\XmlSitemap\RecordsXmlSitemapDataProvider
        config {
          table = tx_googleforjobs_domain_model_job
          sortField = uid
          lastModifiedField = tstamp
          pid = <page id containing job records>
          url {
            pageId = <your detail page id>
            fieldToParameterMap {
              uid = tx_googleforjobs_job[job]
            }
            additionalGetParameters {
              tx_googleforjobs_job.controller = Job
              tx_googleforjobs_job.action = show
            }
            useCacheHash = 1
          }
        }
      }
    }
  }
}
```

### 4) Routing Enhancer

Add the following code to your config.yaml
```
routeEnhancers:
  JobsPlugin:
    type: Extbase
    extension: GoogleForJobs
    plugin: Job
    routes:
      -
        routePath: '/{job_title}'
        _controller: 'Job::show'
        _arguments:
          job_title: job
    defaultController: 'Job::show'
    aspects:
      job_title:
        type: PersistedAliasMapper
        tableName: tx_googleforjobs_domain_model_job
        routeFieldName: path_segment
```

## 3. Administration

### 3.1. Versions and support

| google_for_jobs  | TYPO3      | PHP       | Support/Development                     |
| ---------------- | ---------- | ----------|---------------------------------------- |
| 2.x              | 10.4.x     | 7.2 - 7.4 | Features, Bugfixes, Security Updates    |
| 1.x              | 9.5.x      | 7.0 - 7.2 | Bugfixes, Security Updates              |

#### Update from 1.3 to higher Versions
- Install Update via composer or from TYPO3 TER
- Run the Upgrade Wizard from Admin Tools > Upgrade > Upgrade Wizard. You will be promted to create missing tables and fields, please do so
- Next, execute the "Google For Jobs: migrate locations into seperate table" wizard. This will migrate your existing job locations into a seperate table
- Modify your fluid templates accordingly

### 3.2. Contribution

Pull requests are welcome in general! Nevertheless please don't forget to add an issue and connect it to your pull requests. This
is very helpful to understand what kind of issue the PR is going to solve.

- Bugfixes: Please describe what kind of bug your fix solve and give us feedback how to reproduce the issue. We're going
to accept only bugfixes if I can reproduce the issue.
- Features: Feel free to contact us if u have some cool ideas or wishes.

[1]: https://getcomposer.org/

### 3.3. Changelog

Read the full Changelog [here](./CHANGELOG.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [2.2.0] - 2022-06-16
- compatibility for TYPO3 11

## [2.1.1] - 2021-08-23
Security release: update as soon as possilbe to close a possible XSS vulnerability

### Security
- escape structured data JSON output 
- escape all rendered job properties in fluid templates

## [2.1.0] - 2021-03-08
### Sponsor
- Many thanks to [Cybertrading](https://cybertrading.de/) for sponsoring this update!

### Added
- Support for sytem categories in job records
- Support for list of selected job entries
- Support for setting the order in list view by title, last edited, created and published
- Translation support for Plugin

### Important
- [!!!] Analyze and update database structure after installation

### Fixed
- render missing organization logo url to structured data output
- make validThrough nullable in job model

## [2.0.1] - 2020-06-19
### Fixed
- move flexform registration to tt_content overrides

## [2.0.0] - 2020-06-09
### Added
- Added support for TYPO3 10
- Dropped Support for TYPO3 9

## [1.4.0] - 2020-06-09
### Added
- [!!!][BREAKING] Support for multiple locations per job.
- Support for job locations as seperate records
- Fix: Show job title instead of date in backend record list
- Render structured data to html head as recommended by Google

### Important
- [!!!] Run Upgrade Wizard and update fluid templates. See [here](./README.md#update-from-13-to-higher-Versions)

## [1.3.0] - 2019-11-22
### Added
- Add media, path segment, metadata and alt title to model

### Important
- DB compare needed

## [1.2.0] - 2019-10-28
### Added
- Support for detail view on different page than list view

### Fixed
- correct JSON output for datePosted and validThrough settings

## [1.1.1] - 2019-10-11
### Added
- Changelog.

### Fixed
- Use statement removed from ext_tables.php


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext: "google_for_jobs"
 *
 * Auto generated by Extension Builder 2019-09-19
 *
 * Manual updates:
 * Only the data in the array - anything else is removed by next write.
 * "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = [
    'title' => 'Google for Jobs',
    'description' => 'Render Google for Jobs structured data.',
    'category' => 'plugin',
    'author' => 'Michael Kunst',
    'author_email' => 'kunst@pega-sus.de',
    'author_company' => 'PEGASUS Werbeagentur GmbH',
    'state' => 'stable',
    'uploadfolder' => 0,
    'createDirs' => '',
    'clearCacheOnLoad' => 0,
    'version' => '2.2.0',
    'constraints' => [
        'depends' => [
            'typo3' => '10.0.0-11.9.99',
        ],
        'conflicts' => [],
        'suggests' => [],
    ],
];
